ipn/store/kubestore/store_kube.go

```diff
@@ -84,27 +116,79 @@ func (s *Store) ReadState(id ipn.StateKey) ([]byte, error) {
 
 // WriteState implements the StateStore interface.
 func (s *Store) WriteState(id ipn.StateKey, bs []byte) (err error) {
-	return s.updateStateSecret(map[string][]byte{string(id): bs})
+	defer func() {
+		if err == nil {
+			s.memory.WriteState(ipn.StateKey(sanitizeKey(id)), bs)
+		}
+	}()
+	return s.updateSecret(map[string][]byte{string(id): bs}, s.secretName)
 }
 
-// WriteTLSCertAndKey writes a TLS cert and key to domain.crt, domain.key fields of a Tailscale Kubernetes node's state
-// Secret.
-func (s *Store) WriteTLSCertAndKey(domain string, cert, key []byte) error {
-	return s.updateStateSecret(map[string][]byte{domain + ".crt": cert, domain + ".key": key})
+// WriteTLSCertAndKey writes a TLS cert and key to domain.crt, domain.key fields
+// of a Tailscale Kubernetes node's state Secret.
+func (s *Store) WriteTLSCertAndKey(domain string, cert, key []byte) (err error) {
+	defer func() {
+		if err == nil {
+			s.memory.WriteState(ipn.StateKey(sanitizeKey(domain+".crt")), cert)
```

Comment:
WDYT about using a separate in-memory datastore for certs? We are overlapping the keyspace of multiple Secrets here, which feels like it would make it easy to introduce subtle bugs in future.

Reply:
I was considering that for the previous re-incarnation of the cert sharing where all the certs for ProxyGroup would have come from one Secret and it would been a clear mapping between two stores and two Secrets.
I think at this point, I'd rather avoid changing even more internals unless there is a clear benefit- we should see better how to structure all this once we have BYO certs and perhaps once users start needing to use the read only mode for their own VIPService certs, if they do